src/runtime-rs/crates/hypervisor/src/device_manager.rs

```diff
@@ -0,0 +1,364 @@
+    // subsequent accesses. This index is used to maintain the index at which a
+    // block device is assigned to a container in the sandbox.
+    fn get_and_set_sandbox_block_index(&mut self) -> Result<u64> {
+        let current_index = self.block_index;
+
+        if !self.released_index.is_empty() {
+            match self.released_index.pop() {
+                Some(index) => Ok(index),
+                None => Err(anyhow!("failed to get block index")),
+            }
+        } else {
+            self.block_index += 1;
+            Ok(current_index)
+        }
+    }
+
+    // unsetSandboxBlockIndex deletes the current sandbox block index from BlockIndexMap.
+    // This is used to recover from failure while adding a block device.
+    fn unset_sandbox_block_index(&mut self, index: u64) -> Result<()> {
+        self.released_index.push(index);
+        self.released_index.sort_by(|a, b| b.cmp(a));
+        Ok(())
+    }
+}
+
+pub fn new_device_info(
+    device: &oci::LinuxDevice,
+    bdf: Option<String>,
+    io_limits: Option<IoLimits>,
+) -> Result<GenericConfig> {
+    let allow_device_type: Vec<&str> = vec!["c", "b", "u", "p"];
+
+    info!(sl!(), "linux device info: device path:{:?} ,device type:{:?}, major:{:?}, minor:{:?}, file mode:{:?}, uid:{:?},gid:{:?}", 
+    device.path, device.r#type, device.major, device.minor, device.file_mode, device.uid,device.gid);
+
+    if !allow_device_type.contains(&device.r#type.as_str()) {
+        return Err(anyhow!("runtime not support device type {}", device.r#type));
+    }
+
+    if device.path.is_empty() {
+        return Err(anyhow!("container path can not be empty"));
+    }
+
+    let file_mode = device.file_mode.unwrap_or(0);
+    let uid = device.uid.unwrap_or(0);
+    let gid = device.gid.unwrap_or(0);
+
+    let dev_info = GenericConfig {
+        host_path: String::new(),
+        container_path: device.path.clone(),
+        dev_type: device.r#type.clone(),
+        major: device.major,
+        minor: device.minor,
+        file_mode: file_mode as u32,
+        uid,
+        gid,
+        id: "".to_string(),
+        bdf,
+        driver_options: HashMap::new(),
+        io_limits,
+        ..Default::default()
+    };
+    Ok(dev_info)
+}
+
+fn is_block(dev_info: &GenericConfig) -> bool {
+    dev_info.dev_type == "b"
+}
+
+// get_host_path is used to fetch the host path for the device.
+// The path passed in the spec refers to the path that should appear inside the container.
+// We need to find the actual device path on the host based on the major-minor numbers of the device.
+fn get_host_path(dev_info: &GenericConfig) -> Result<String> {
+    if dev_info.container_path.is_empty() {
+        return Err(anyhow!("Empyt path provided for device"));
+    }
+
+    let path_comp = match dev_info.dev_type.as_str() {
+        "c" | "u" => "char",
+        "b" => "block",
+        _ => return Ok(String::new()),
```

Comment:
This is for `"p"` right?

Reply:
yes

Reply:
Suggest adding some comments about in which case returning an empty string.

Reply:
fixed